### PR TITLE
Endpoint validation

### PR DESCRIPTION
**Pull request by rthallisey**
Validate we are connected to endpoints before we start the broker.

### CODE REVIEW

File: pkg/app/client.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package app
+
+import "os/exec"
+
+func GetCOEVersion() ([]byte, error) {
```
Comment (jmrodri) [inline pkg/app/client.go]: What does COE mean?

File: pkg/dao/dao.go [hunk 2]
```diff
@@ -63,6 +67,34 @@ func (d *Dao) SetRaw(key string, val string) error {
...
+		}
+		return vresp.Server, vresp.Cluster, nil
+	default:
+		var connectErr error
+		if err := json.Unmarshal(body, &connectErr); err != nil {
+			return "", "", err
+		}
+		return "", "", connectErr
+	}
+}
```
Comment (jmrodri) [inline pkg/dao/dao.go]: +1

File: pkg/app/client.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+	return runCommand("oc", "version")
+}
+
+func runCommand(cmd string, args ...string) ([]byte, error) {
+	output, err := exec.Command(cmd, args...).CombinedOutput()
+	return output, err
+}
```
Comment (jmrodri) [inline pkg/app/client.go]: I'll allow it, we should probably put all the runCommands in a common util package somewhere for both broker and apb packages to use.

File: pkg/dao/dao.go (hunk 2 above)
Comment (eriknelson) [inline pkg/dao/dao.go]: This is cool; so we're just confirming a connection here, not validating a version range?

File: pkg/app/client.go (hunk 3 above)
Comment (eriknelson) [inline pkg/app/client.go]: We actually already have this segregated in `pkg/apb/hack.go`. Could make the function public and access it via `apb.RunCommand`, but that's not a semantically appropriate place for it.

I'd say we could put it in the `app` pkg, but we'll run into circular dependency problems with `app` importing `apb`, and `apb` importing `app`. It's a problem I've seen in a number of cases.

@jmrodri +1 on the utils pkg, although is it bad form to continue to add to our already sizable pkg count?

File: pkg/app/client.go (hunk 1 above)
Comment (rthallisey) [inline pkg/app/client.go]: COE is ContainerOrchestrationEngine.  I was thinking of first trying with the OpenShift client and if that fails try kubectl.  That way we can support both clients.

File: pkg/app/client.go (hunk 3 above)
Comment (rthallisey) [inline pkg/app/client.go]: Agreed.

File: pkg/dao/dao.go (hunk 2 above)
Comment (rthallisey) [inline pkg/dao/dao.go]: Ya this confirms connection.  Later we could always check for version issues.

File: pkg/app/client.go (hunk 3 above)
Comment (rthallisey) [inline pkg/app/client.go]: I think a utils pkg at the top level is a good idea.  I created https://github.com/fusor/ansible-service-broker/issues/62 to track that.  I think we should have a separate patch to fix that.

Comment (eriknelson) [inline pkg/app/client.go]: SGTM

**Comment by jmrodri** [comment]
TEST
-------

1. bring up broker without etcd, expect gracious error message
2. bring up broker with etcd, ensure broker functions normally

**Comment by jmrodri** [comment]
TEST RESULTS
-----------------------

1. bring up broker without etcd, expect gracious error message

```
[jesusr@speed3 ansible-service-broker{master}]$ make run
# HACK: Unless docker's vendor directory is removed, we end up with a
# duplicate symbol error from the linker that prevents compilation.
rm -rf /home/jesusr/dev/src/github.com/fusor/ansible-service-broker/vendor/github.com/docker/docker/vendor && \
	go install ./cmd/broker
============================================================
==           Starting Ansible Service Broker...           ==
============================================================
[2017-05-03T14:00:49.91-04:00] [DEBUG] Connecting Dao
[2017-05-03T14:00:49.91-04:00] [INFO] == ETCD CX ==
[2017-05-03T14:00:49.91-04:00] [INFO] EtcdHost: localhost
[2017-05-03T14:00:49.91-04:00] [INFO] EtcdPort: 2379
[2017-05-03T14:00:49.91-04:00] [INFO] Endpoints: [http://localhost:2379]
[2017-05-03T14:00:49.91-04:00] [ERROR] Failed to connect to Etcd

[2017-05-03T14:00:49.91-04:00] [ERROR] Get http://localhost:2379/version: dial tcp [::1]:2379: getsockopt: connection refused
Makefile:17: recipe for target 'run' failed
make: *** [run] Error 1
```
We get a nice error message (in red) that we failed to connect to Etcd. Before we would start and fail only after bootstrapping which is way to late.

**Comment by jmrodri** [comment]
TEST RESULTS
------------------------
2. bring up broker with etcd, but NO OPENSHIFT CLUSTER

Fails with nasty panic:

```

[jesusr@speed3 ansible-service-broker{master}]$ make run
# HACK: Unless docker's vendor directory is removed, we end up with a
# duplicate symbol error from the linker that prevents compilation.
rm -rf /home/jesusr/dev/src/github.com/fusor/ansible-service-broker/vendor/github.com/docker/docker/vendor && \
	go install ./cmd/broker
============================================================
==           Starting Ansible Service Broker...           ==
============================================================
[2017-05-03T14:03:19.864-04:00] [DEBUG] Connecting Dao
[2017-05-03T14:03:19.864-04:00] [INFO] == ETCD CX ==
[2017-05-03T14:03:19.864-04:00] [INFO] EtcdHost: localhost
[2017-05-03T14:03:19.864-04:00] [INFO] EtcdPort: 2379
[2017-05-03T14:03:19.864-04:00] [INFO] Endpoints: [http://localhost:2379]
[2017-05-03T14:03:19.866-04:00] [INFO] Etcd Version [Server: 3.1.3, Cluster: 3.1.0]
[2017-05-03T14:03:19.866-04:00] [DEBUG] Connecting Registry
[2017-05-03T14:03:19.866-04:00] [INFO] == REGISTRY CX == 
[2017-05-03T14:03:19.866-04:00] [INFO] Name: dockerhub
[2017-05-03T14:03:19.866-04:00] [INFO] Url: http://localhost:1337
[2017-05-03T14:03:19.866-04:00] [DEBUG] DockerHubRegistry::Init
[2017-05-03T14:03:19.866-04:00] [DEBUG] Connecting to OpenShift
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0xce1b20, 0xc420012180)
	/usr/lib/golang/src/runtime/panic.go:500 +0x1a1
github.com/fusor/ansible-service-broker/pkg/app.CreateApp(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/jesusr/dev/src/github.com/fusor/ansible-service-broker/pkg/app/app.go:100 +0x11db
main.main()
	/home/jesusr/dev/src/github.com/fusor/ansible-service-broker/cmd/broker/main.go:6 +0x33
Makefile:17: recipe for target 'run' failed
make: *** [run] Error 2
```

I put in a log statement in app.go to printout the output.

```
[2017-05-03T14:08:07.377-04:00] [DEBUG] Connecting to OpenShift
[2017-05-03T14:08:07.485-04:00] [NOTICE] [oc v3.6.0-alpha.1+46942ad kubernetes v1.5.2+43a9be4 features: Basic-Auth GSSAPI Kerberos SPNEGO ]
```

**Comment by jmrodri** [comment]
Added more debug.

```
    ver, err := GetCOEVersion()
    app.log.Notice(string(ver))
```
Outputs the following 3 lines of output:
```
[2017-05-03T14:12:58.352-04:00] [NOTICE] oc v3.6.0-alpha.1+46942ad
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO
```

if I loop through the output array after splitting it:

```
    output := strings.Split(string(ver), "\n")

    app.log.Notice(fmt.Sprintf("%v", output))
    for s := range output {
        app.log.Notice(output[s])
    }
```

 we get the following.

```
[2017-05-03T14:12:58.352-04:00] [NOTICE] [oc v3.6.0-alpha.1+46942ad kubernetes v1.5.2+43a9be4 features: Basic-Auth GSSAPI Kerberos SPNEGO ]
[2017-05-03T14:12:58.352-04:00] [NOTICE] oc v3.6.0-alpha.1+46942ad
[2017-05-03T14:12:58.352-04:00] [NOTICE] kubernetes v1.5.2+43a9be4
[2017-05-03T14:12:58.352-04:00] [NOTICE] features: Basic-Auth GSSAPI Kerberos SPNEGO
[2017-05-03T14:12:58.352-04:00] [NOTICE] 
```

The problem resides in this loop:

```
    // Output only the response from the server
    //   Server https://xxx.xx.xx.xx:yyyy
    //   openshift vx.y.z
    //   kubernetes vx.y.z
    for o := 4; o < 7; o++ {
        app.log.Info(output[o])
    }
```
Those indexes don't work.

**Comment by jmrodri** [comment]
```
[jesusr@speed3 ansible-service-broker{master}]$ oc version
oc v3.6.0-alpha.1+46942ad
kubernetes v1.5.2+43a9be4
features: Basic-Auth GSSAPI Kerberos SPNEGO
```

**Comment by rthallisey** [comment]
I'm going to change this so it doesn't rely on the client.  Instead we'll make a rest request.

**Comment by jmrodri** [comment]
@rthallisey rest sounds good.

**Comment by rthallisey** [comment]
I want to create a Kubernetes and Openshift client, but I'm having dependency issues. Fixing https://github.com/fusor/ansible-service-broker/issues/100 first.

**Comment by rthallisey** [comment]
Update: After I got around the dependency issues I was having, I dug deeper into how we could run this validation by using a kubernetes client object. If we use a kubernetes client object, we'll have to read from a config file.  There are two places where we run the broker, locally (make run) and as a pod.  Getting the configfile as a pod is easy to handle, but locally we'll have to read from $HOME/.kube/confg.   Reading from a local config isn't ideal.  I'll see if I can come up with another way to do this.

**Comment by jmrodri** [comment]
@rthallisey can we not just use a normal http request and hit the configured address?Like `curl -X GET http://IPADDR:PORT/SOME/ENDPOINT`

**Comment by rthallisey** [comment]
That's true. Since I'm creating a rest client might as well skip that and just do a request to the endpoint.

**Comment by rthallisey** [comment]
For a normal http request we will still need a token to contact kubernetes. So there's no difference between that and creating a client because we're going to have to get that from the config.

**Comment by rthallisey** [comment]
ref #103

**Comment by rthallisey** [comment]
I removed the cluster validation which depends on the completion of #103.

**Comment by jmrodri** [comment]
VERIFIED!

### etcd down
```
[jesusr@speed3 ansible-service-broker{rthallisey-endpoint-validation}]$ make run                                            
# HACK: Unless docker's vendor directory is removed, we end up with a                                                       
# duplicate symbol error from the linker that prevents compilation.                                                         
rm -rf /home/jesusr/dev/src/github.com/fusor/ansible-service-broker/vendor/github.com/docker/docker/vendor && \             
        go install ./cmd/broker                                                                                             
============================================================                                                                
==           Starting Ansible Service Broker...           ==                                                                
============================================================                                                                
[2017-05-15T20:45:31.333-04:00] [DEBUG] Connecting Dao                                                                      
[2017-05-15T20:45:31.333-04:00] [INFO] == ETCD CX ==                                                                        
[2017-05-15T20:45:31.333-04:00] [INFO] EtcdHost: localhost                                                                  
[2017-05-15T20:45:31.333-04:00] [INFO] EtcdPort: 2379                                                                       
[2017-05-15T20:45:31.333-04:00] [INFO] Endpoints: [http://localhost:2379]                                                   
[2017-05-15T20:45:31.333-04:00] [ERROR] Failed to connect to Etcd                                                           
                                                                                                                            
[2017-05-15T20:45:31.333-04:00] [ERROR] Get http://localhost:2379/version: dial tcp [::1]:2379: getsockopt: connection refus
ed                                                                                                                          
Makefile:17: recipe for target 'run' failed                                                                                 
make: *** [run] Error 1            
```
### etcd up

```                                                                                         
[jesusr@speed3 ansible-service-broker{rthallisey-endpoint-validation}]$ make run                                            
# HACK: Unless docker's vendor directory is removed, we end up with a                                                       
# duplicate symbol error from the linker that prevents compilation.                                                         
rm -rf /home/jesusr/dev/src/github.com/fusor/ansible-service-broker/vendor/github.com/docker/docker/vendor && \             
        go install ./cmd/broker                                                                                             
============================================================                                                                
==           Starting Ansible Service Broker...           ==                                                                
============================================================                                                                
[2017-05-15T20:46:03.938-04:00] [DEBUG] Connecting Dao                                                                      
[2017-05-15T20:46:03.938-04:00] [INFO] == ETCD CX ==                                                                        
[2017-05-15T20:46:03.938-04:00] [INFO] EtcdHost: localhost                                                                  
[2017-05-15T20:46:03.938-04:00] [INFO] EtcdPort: 2379                                                                       
[2017-05-15T20:46:03.938-04:00] [INFO] Endpoints: [http://localhost:2379]                                                   
[2017-05-15T20:46:03.939-04:00] [INFO] Etcd Version [Server: 3.1.3, Cluster: 3.1.0]                                         
[2017-05-15T20:46:03.939-04:00] [DEBUG] Connecting Registry                                                                 
[2017-05-15T20:46:03.939-04:00] [INFO] == REGISTRY CX ==                                                                    
[2017-05-15T20:46:03.939-04:00] [INFO] Name: dockerhub                                                                      
[2017-05-15T20:46:03.939-04:00] [INFO] Url: http://localhost:1337                                                           
[2017-05-15T20:46:03.939-04:00] [DEBUG] DockerHubRegistry::Init                                                             
[2017-05-15T20:46:03.939-04:00] [DEBUG] Initializing WorkEngine                                                             
[2017-05-15T20:46:03.939-04:00] [DEBUG] Initializing Provision WorkSubscriber                                               
[2017-05-15T20:46:03.939-04:00] [DEBUG] Creating AnsibleBroker                                                              
[2017-05-15T20:46:03.939-04:00] [NOTICE] Ansible Service Broker Started                                                     
[2017-05-15T20:46:03.939-04:00] [NOTICE] Listening on http://0.0.0.0:1338                                                   
[2017-05-15T20:46:03.939-04:00] [INFO] Listening for provision messages                                                     
```

**Pull request merged**